File: typescript/src/01-strategy/before/payment-service.ts
import type { PaymentResult } from '../../shared/models.js';

// Classic if-else jungle: every new payment method means editing this file.
// In production this pattern grows past 200 lines; the demo file stays short.

const REQUIRED = ['UPI', 'Card', 'Wallet', 'NetBanking', 'EMI', 'PhonePe'];

// Simulates the prod vault. Set SIMULATE_BLAST_RADIUS=1 to replay the Friday incident.
function loadAllKeys(): Record<string, string> {
  const simulateMissingPhonePe = process.env.SIMULATE_BLAST_RADIUS === '1';
  const keys: Record<string, string> = {};
  for (const method of REQUIRED) {
    if (simulateMissingPhonePe && method === 'PhonePe')
      throw new Error('Missing vault credential: PhonePe (prod deploy forgot to add key)');
    keys[method] = `sk-${method.toLowerCase()}-demo-key`;
  }
  return keys;
}

// Shared config - all providers load keys from one block at MODULE LOAD time.
//    This is the Node/TS analog of C#'s static constructor: if loadAllKeys() throws,
//    importing this module fails, so EVERY payment method goes down - not just PhonePe.
//    (See README "Blast radius across runtimes".)
const GATEWAY_KEYS: Record<string, string> = loadAllKeys();
console.log('  [Before] Gateway keys loaded from vault (shared init OK)');

export class PaymentService {
  process(method: string, amount: number): PaymentResult {
    const apiKey = GATEWAY_KEYS[method]; // every branch depends on the shared dictionary
    if (apiKey === undefined) throw new Error(`Unknown payment method: ${method}`);

    // In production, imagine ten more else-ifs. Each new provider adds a branch here.
    if (method === 'UPI') {
      console.log(`  [Before/UPI] Collect ₹${amount} | VPA verify | key ${apiKey.slice(0, 8)}...`);
      console.log('  [Before/UPI] NPCI rail OK -> merchant credited');
      return { success: true, message: 'UPI collect initiated' };
    } else if (method === 'Card') {
      console.log(`  [Before/Card] Auth hold ₹${amount} | token ${apiKey.slice(0, 8)}... | 3DS=pending`);
      return { success: true, message: 'Card authorized (3DS redirect queued)' };
    } else if (method === 'Wallet') {
      console.log(`  [Before/Wallet] Debit wallet ₹${amount} | balance check | sk=${apiKey.slice(0, 6)}...`);
      return { success: true, message: 'Wallet balance debited' };
    } else if (method === 'NetBanking') {
      console.log(`  [Before] NetBanking redirect | amt=₹${amount} | session key ${apiKey.slice(0, 8)}...`);
      return { success: true, message: 'NetBanking redirect URL generated' };
    } else if (method === 'EMI') {
      console.log(`  [Before/EMI] Tenure split on ₹${amount} | rate lookup | cred ${apiKey.slice(0, 8)}...`);
      return { success: true, message: 'EMI plan attached to order' };
    } else if (method === 'PhonePe') {
      console.log(`  [Before/PhonePe] Intent ₹${amount} | merchantId=${apiKey.slice(0, 8)}... | deep-link pending`);
      return { success: true, message: 'PhonePe payment intent created' };
    }
    throw new Error(`Unknown payment method: ${method}`);
  }
}

File: csharp/DesiArchitect.DesignPatterns/01_Strategy/Before/PaymentService.cs
using DesiArchitect.DesignPatterns.Shared;

namespace DesiArchitect.DesignPatterns.Strategy.Before;

// Classic if-else jungle: every new payment method means editing this file.
// In production this pattern grows past 200 lines; the demo file stays short.
public class PaymentService
{
    // All providers load keys from one shared block at startup.
    private static readonly IReadOnlyDictionary<string, string> GatewayKeys;

    static PaymentService()
    {
        GatewayKeys = GatewayVault.LoadAllKeys();
        Console.WriteLine("  [Before] Gateway keys loaded from vault (shared init OK)");
    }

    public PaymentResult Process(string method, decimal amount)
    {
        var apiKey = GatewayKeys[method];   // every branch depends on the shared dictionary

        if (method == "UPI")
        {
            Console.WriteLine($"  [Before/UPI] Collect ₹{amount} | VPA verify | key {apiKey[..8]}...");
            Console.WriteLine("  [Before/UPI] NPCI rail OK -> merchant credited");
            return new PaymentResult(true, "UPI collect initiated");
        }
        else if (method == "Card")
        {
            Console.WriteLine($"  [Before/Card] Auth hold ₹{amount} | token {apiKey[..8]}... | 3DS=pending");
            return new PaymentResult(true, "Card authorized (3DS redirect queued)");
        }
        else if (method == "Wallet")
        {
            Console.WriteLine($"  [Before/Wallet] Debit wallet ₹{amount} | balance check | sk={apiKey[..6]}...");
            return new PaymentResult(true, "Wallet balance debited");
        }
        else if (method == "NetBanking")
        {
            Console.WriteLine($"  [Before] NetBanking redirect | amt=₹{amount} | session key {apiKey[..8]}...");
            return new PaymentResult(true, "NetBanking redirect URL generated");
        }
        else if (method == "EMI")
        {
            Console.WriteLine($"  [Before/EMI] Tenure split on ₹{amount} | rate lookup | cred {apiKey[..8]}...");
            return new PaymentResult(true, "EMI plan attached to order");
        }
        else if (method == "PhonePe")
        {
            Console.WriteLine($"  [Before/PhonePe] Intent ₹{amount} | merchantId={apiKey[..8]}... | deep-link pending");
            return new PaymentResult(true, "PhonePe payment intent created");
        }
        // In production, imagine ten more else-ifs. Each new provider adds a branch and vault key here.
        else
        {
            throw new ArgumentException($"Unknown payment method: {method}");
        }
    }

}

// Simulates prod vault - set SIMULATE_BLAST_RADIUS=1 to replay Friday incident locally
internal static class GatewayVault
{
    private static readonly string[] Required =
        ["UPI", "Card", "Wallet", "NetBanking", "EMI", "PhonePe"];

    public static Dictionary<string, string> LoadAllKeys()
    {
        var simulateMissingPhonePe = Environment.GetEnvironmentVariable("SIMULATE_BLAST_RADIUS") == "1";
        var keys = new Dictionary<string, string>();

        foreach (var method in Required)
        {
            if (simulateMissingPhonePe && method == "PhonePe")
                throw new InvalidOperationException("Missing vault credential: PhonePe (prod deploy forgot to add key)");

            keys[method] = $"sk-{method.ToLowerInvariant()}-demo-key";
        }

        return keys;
    }
}